boards/BMS/Src/Io/Io_Eeprom.c

```diff
@@ -0,0 +1,154 @@
+#include "Io_Eeprom.h"
+
+// Define the I2C
+extern I2C_HandleTypeDef hi2c1;
+#define EEPROM_I2C &hi2c1
+
+// EEPROM ADDRESS (8bits)
+#define EEPROM_ADDR 0xA0
+
+// Define the Page Size and number of pages
+#define PAGE_SIZE ((uint16_t)16U)                            // in Bytes
+#define PAGE_NUM ((uint16_t)1024U)                           // number of pages
+#define PAGE_ADDR_START_BIT ((int)(log(PAGE_SIZE) / log(2))) // number of bit where addressing starts
+
+// Function to determine the number of bytes to write/read to each page.
+uint16_t Bytes_To_Process(uint16_t size, uint16_t offset)
+{
+    if ((size + offset) < PAGE_SIZE)
+    {
+        return size;
+    }
+    else
+    {
+        return (uint16_t)(PAGE_SIZE - offset);
+    }
+}
+
+void convert_float_to_bytes(uint8_t *bytes, float float_to_convert)
+{
+    // Create union that stores float and byte array in same memory location.
+    // This allows you to access 8-bit segments of the float value using array indexing
+    union
+    {
+        float   float_val;
+        uint8_t bytes[4];
+    } u;
+
+    // place float input into union
+    u.float_val = float_to_convert;
+
+    for (int i = 0; i < 4; i++)
+    {
+        // convert to array of bytes by accessing float value in union in byte-size increments (pun-intended)
+        bytes[i] = u.bytes[i];
+    }
+}
+
+float convert_bytes_to_float(uint8_t *bytes)
+{
+    // Create union that stores float and byte array in same memory location.
+    // This allows you to access 8-bit segments of the float value using array indexing
+    union
+    {
+        float   float_val;
+        uint8_t bytes[4];
+    } u;
+
+    for (int i = 0; i < 4; i++)
+    {
+        u.bytes[i] = bytes[i];
+    }
+
+    return u.float_val;
+}
+
+void Io_Eeprom_Write(uint16_t page, uint16_t offset, uint8_t *data, uint16_t size)
+{
+    // determine the start page and ending page
+    uint16_t startPage = page;
+    uint16_t endPage   = (uint16_t)(page + ((size + offset) / PAGE_SIZE));
+
+    // determine the number of pages to be written
+    uint16_t num_pages = (uint16_t)(endPage - startPage + 1);
+    uint16_t pos       = 0U;
+
+    for (int i = 0; i < num_pages; i++)
+    {
+        // determine the memory address location
+        uint16_t mem_address = (uint16_t)(startPage << PAGE_ADDR_START_BIT | offset);
+
+        // Determine the number of bytes to write to page
+        uint16_t bytes_to_write = Bytes_To_Process(size, offset);
+
+        HAL_I2C_Mem_Write(EEPROM_I2C, EEPROM_ADDR, mem_address, 2, &data[pos], bytes_to_write, 1000);
+
+        startPage++;                                // Increment the page
+        offset = 0;                                 // writing to new page, offset will be zero for next write.
+        size   = (uint16_t)(size - bytes_to_write); // reduce size by the amount of bytes written to page
+        pos    = (uint16_t)(pos + bytes_to_write);  // increase starting position by number of bytes written
```

Comment:
tried to follow how this is iterating as you're accessing `data[] = bytes_to_write[]` with size `4` at position `pos`, but you appear to be incrementing `pos` beyond size `4` (buffer overread?). some recommendations to improve readability here:

replace `convert_float_to_bytes()` with:
```c
float f = ...;
const uint8_t const * data = (const uint8_t const *)&f; // const pointer to const int
```

simplify `convert_bytes_to_float()` with:
```c
memcpy(&f, data, sizeof(f));
```

this should allow you to remove the following functions entirely: `Io_Eeprom_WriteFloat()`, `Io_Eeprom_ReadFloat()`, `convert_float_to_bytes()`, `convert_bytes_to_float()`

P.S.
- define magic numbers like `4` in one place to improve maintainability
- remember to scope local functions with `static` and not expose them to the rest of the program

Reply:
Definitely some good recommendations here, I'll be sure to apply them

As for the flow, I'll lay it out best I can and that should either clarify or exposes flaws in my design:

1. `data[]` is an array of bytes to be written to the EEPROM, it can be of any size (theoretically I suppose it should be limited to at least the size of the EEPROM)
2. The EEPROM can only be written to one page at a time, each page being 16 bits wide (4 bytes), `num_pages` determines the number of individual page writes necessary to write contents of `data[]` to the EEPROM, and is dependant on the starting `page`, `offset` as well as the total number of bytes to be written, `size`
3. `Bytes_To_Process` returns the number of bytes to be read or written to the current page, in this case `bytes_to_write` contains the number of bytes to write to the current page.
4. ` HAL_I2C_Mem_Write` writes `bytes_to_write` number of bytes to the current page, starting at `mem_address`
5. `pos` (the index of the input data array) is incremented by the number of bytes written (between 1-4)
6. for loop continues until we have written to the necessary number of pages to write entirety of `data[]` to the EEPROM

I think that's the important bits at least... obviously I could do better to clean up and comment to make this more clear. It should be noted that this has not yet been tested, and so there very well could be bugs in my code. Hopefully this clears up at least a little of the confusion as to my intentions at least

Reply:
thanks for the explanation 👍. i may be misunderstanding but my main concern was passing `bytes_to_write[]` of size 4 to `Io_Eeprom_Write()` as `data[]` and accessing that at `pos` which is being incremented beyond size 4 (?)

```c
void Io_Eeprom_WriteFloat(uint16_t page, uint16_t offset, float data)
...
    uint8_t bytes_to_write[4];
...
    Io_Eeprom_Write(page, offset, bytes_to_write, 4);
...
void Io_Eeprom_Write(uint16_t page, uint16_t offset, uint8_t *data, uint16_t size)
...
        HAL_I2C_Mem_Read(EEPROM_I2C, EEPROM_ADDR, mem_address, 2, &data[pos], bytes_to_read, 1000);
...
        pos    = (uint16_t)(pos + bytes_to_read);  // increase starting position by number of bytes read
```

if i'm misunderstanding, a check that `pos < size` would still help. anyways probs only worth refactoring after clean up + testing

Reply:
You can use `sizeof(float)` too instead of 4 since that indicates what 4 represents in this case

Reply:
Oh I see the confusion, you're referencing `Io_Eeprom_WriteFloat` `bytes_to_write`, yeah that's a terrible variable name, my bad lol.

As for `pos`, in the case you have above, `Io_Eeprom_WriteFloat` writes a single 4-byte float value to the EEPROM, because this would fit inside a single page (16 bytes) the for loop inside `Io_Eeprom_Write` would only run for a single cycle, so while `pos` is updated, the updated value is never used, as `num_pages` is 1 and the for loop will break. 

if `num_pages` was larger than 1, (say writing 20 bytes to the EEPROM), then the updated value of `pos` would indicate the starting index in `data` to write to the second page.

So yes, `pos` does go greater than `size`, but the for loop will break under these conditions. An extra sanity check could be added, but I don't think it would ever be exercised.